File: sample/src/main/java/com/cabe/lib/ui/sample/MainActivity.java
package com.cabe.lib.ui.sample;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.cabe.lib.ui.widget.LoadMoreRecyclerViewX;
import com.cabe.lib.ui.widget.OnEndViewListener;
import com.cabe.lib.ui.widget.OnLoadViewListener;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

/**
 * 作者：沈建芳 on 2019-05-16 16:01
 */
public class MainActivity extends AppCompatActivity {
    private LoadMoreRecyclerViewX recyclerView;
    private MyAdapter mMyAdapter = new MyAdapter();
    private int curPageIndex = 0;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = findViewById(R.id.activity_main_recycler);
        recyclerView.setAdapter(mMyAdapter);
        recyclerView.setScrollCallback(() -> recyclerView.postDelayed(() -> appendData(++ curPageIndex), 1000));
        recyclerView.postDelayed(this::loadData, 1000);

        recyclerView.setOnLoadViewListener(new OnLoadViewListener() {
            @Override
            public View onCreateLoadView(ViewGroup parent) {
                return null;
            }
            @Override
            public void onLoadViewBind(View loadView) {
                loadView.setBackgroundColor(0xFFFF0000);
                TextView label = (TextView) ((ViewGroup) loadView).getChildAt(1);
                label.setTextColor(0xFFFFFFFF);
            }
        });
        recyclerView.setOnEndViewListener(new OnEndViewListener() {
            @Override
            public View onCreateEndView(ViewGroup parent) {
                return null;
            }
            @Override
            public void onEndViewBind(View loadView) {
                loadView.setBackgroundColor(0xFF0000FF);
                TextView label = (TextView) ((ViewGroup) loadView).getChildAt(0);
                label.setTextColor(0xFF00FFFF);
            }
        });

        SwipeRefreshLayout swipeLayout = findViewById(R.id.activity_main_swipe);
        swipeLayout.setOnRefreshListener(() -> recyclerView.postDelayed(() -> {
            swipeLayout.setRefreshing(false);
            loadData();
        }, 1000));
    }

    private void loadData() {
        curPageIndex = 0;
        List<String> dataList = new ArrayList<>();
        for(int i=0;i<10;i++) {
            dataList.add("loadData_" + i);
        }
        mMyAdapter.setDataList(dataList);
    }

    private void appendData(int pageIndex) {
        List<String> dataList = new ArrayList<>();
        for(int i=0;i<10;i++) {
            dataList.add("appendData#" + pageIndex + "_" + i);
        }
        mMyAdapter.appendDataList(dataList);
        if(pageIndex > 1) {
            recyclerView.setScrollEnd(true);
        }
    }

    private class MyAdapter extends RecyclerView.Adapter<MyViewHolder> {
        private List<String> dataList = new ArrayList<>();
        private void setDataList(List<String> list) {
            dataList = list;
            notifyDataSetChanged();
        }
        private void appendDataList(List<String> list) {
            int count = getItemCount();
            if(list != null) {
                dataList.addAll(list);
            }
            notifyItemInserted(count);
        }
        private String getItemData(int position) {
            if(position < 0 || position >= getItemCount()) return null;

            return dataList.get(position);
        }
        @Override
        public int getItemCount() {
            return dataList == null ? 0 : dataList.size();
        }
        @Override
        public MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_layout_view, parent, false);
            return new MyViewHolder(itemView);
        }

        @Override
        public void onBindViewHolder(MyViewHolder holder, int position) {
            String data = getItemData(position);
            holder.rowLabel.setText(data);
        }
    }

    private class MyViewHolder extends RecyclerView.ViewHolder {
        private TextView rowLabel;
        MyViewHolder(View itemView) {
            super(itemView);
            rowLabel = itemView.findViewById(R.id.item_layout_view);
        }
    }
}

File: sample/src/main/java/com/cabe/lib/ui/sample/MainActivity.kt
package com.cabe.lib.ui.sample

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.cabe.lib.ui.decoration.FastIndexByLetterDecoration
import com.cabe.lib.ui.decoration.FloatButtonDecoration
import com.cabe.lib.ui.decoration.HorizontalScrollBarDecoration
import com.cabe.lib.ui.widget.OnEndViewListener
import com.cabe.lib.ui.widget.OnLoadViewListener
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

/**
 * 作者：沈建芳 on 2019-05-16 16:01
 */
class MainActivity : AppCompatActivity() {
    private val mMyAdapter = MyAdapter()
    private var curPageIndex = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        activity_main_recycler_h.adapter = MyAdapterH().apply {
            setDataList(arrayListOf<String>().apply {
                for(i in 0..5) {
                    add(i.toString())
                }
            })
        }
        activity_main_recycler_h.setAutoLoad(false)
        activity_main_recycler_h.setShowEnd(false)
        activity_main_recycler_h.addItemDecoration(HorizontalScrollBarDecoration())

        activity_main_recycler.adapter = mMyAdapter
        activity_main_recycler.setScrollCallback { activity_main_recycler.postDelayed({ appendData(++curPageIndex) }, 1000) }
        activity_main_recycler.postDelayed({ loadData() }, 1000)
        activity_main_recycler.setOnLoadViewListener(object : OnLoadViewListener {
            override fun onCreateLoadView(parent: ViewGroup): View? {
                return null
            }
            override fun onLoadViewBind(loadView: View) {
                loadView.setBackgroundColor(-0x10000)
                val label = (loadView as ViewGroup).getChildAt(1) as TextView
                label.setTextColor(-0x1)
            }
        })
        activity_main_recycler.setOnEndViewListener(object : OnEndViewListener {
            override fun onCreateEndView(parent: ViewGroup): View? {
                return null
            }
            override fun onEndViewBind(loadView: View) {
                loadView.setBackgroundColor(-0xffff01)
                val label = (loadView as ViewGroup).getChildAt(0) as TextView
                label.setTextColor(-0xff0001)
            }
        })
        activity_main_recycler.addItemDecoration(FloatButtonDecoration())
        activity_main_recycler.addItemDecoration(FastIndexByLetterDecoration {

        })
        activity_main_swipe.setOnRefreshListener {
            activity_main_recycler.postDelayed({
                activity_main_swipe.isRefreshing = false
                loadData()
            }, 1000)
        }
    }

    private fun loadData() {
        curPageIndex = 0
        val dataList: MutableList<String> = ArrayList()
        for (i in 0..9) {
            dataList.add("loadData_$i")
        }
        mMyAdapter.setDataList(dataList)
    }

    private fun appendData(pageIndex: Int) {
        val dataList: MutableList<String> = ArrayList()
        for (i in 0..9) {
            dataList.add("appendData#" + pageIndex + "_" + i)
        }
        mMyAdapter.appendDataList(dataList)
        if (pageIndex > 1) {
            activity_main_recycler!!.setScrollEnd(true)
        }
    }

    private inner class MyAdapter : RecyclerView.Adapter<MyViewHolder>() {
        private var dataList: MutableList<String>? = ArrayList()
        fun setDataList(list: MutableList<String>) {
            dataList = list
            notifyDataSetChanged()
        }

        fun appendDataList(list: List<String>?) {
            val count = itemCount
            if (list != null) {
                dataList!!.addAll(list)
            }
            notifyItemInserted(count)
        }

        private fun getItemData(position: Int): String? {
            return if (position < 0 || position >= itemCount) null else dataList!![position]
        }

        override fun getItemCount(): Int {
            return if (dataList == null) 0 else dataList!!.size
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
            val itemView = LayoutInflater.from(parent.context).inflate(R.layout.item_layout_view, parent, false)
            return MyViewHolder(itemView)
        }

        override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
            val data = getItemData(position)
            holder.rowLabel.text = data
        }
    }

    private inner class MyViewHolder(itemView: View) : ViewHolder(itemView) {
        val rowLabel: TextView = itemView.findViewById(R.id.item_layout_view)
    }

    private inner class MyAdapterH : RecyclerView.Adapter<VH>() {
        private var dataList: MutableList<String>? = ArrayList()
        fun setDataList(list: MutableList<String>) {
            dataList = list
            notifyDataSetChanged()
        }

        private fun getItemData(position: Int): String? {
            return if (position < 0 || position >= itemCount) null else dataList!![position]
        }

        override fun getItemCount(): Int {
            return if (dataList == null) 0 else dataList!!.size
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
            val itemView = LayoutInflater.from(parent.context).inflate(R.layout.item_layout_view_h, parent, false)
            return VH(itemView)
        }

        override fun onBindViewHolder(holder: VH, position: Int) {
            val data = getItemData(position)
            holder.rowLabel.text = data
        }
    }
    private inner class VH(itemView: View) : ViewHolder(itemView) {
        val rowLabel: TextView = itemView.findViewById(R.id.item_layout_view)
    }
}